File: sdk/typescript/src/namespaces/expenses.ts
/**
 * Expenses Namespace API
 *
 * Provides a namespaced interface for expense tracking and management.
 * Supports receipt processing, auto-categorization, approval workflows, and QBO sync.
 */
import type { PaginationParams } from '../types';
// =============================================================================
// Types
// =============================================================================

/** Expense categories */
export type ExpenseCategory =
  | 'travel'
  | 'meals'
  | 'office_supplies'
  | 'software'
  | 'equipment'
  | 'professional_services'
  | 'marketing'
  | 'utilities'
  | 'rent'
  | 'insurance'
  | 'other';

/** Expense status */
export type ExpenseStatus = 'pending' | 'approved' | 'rejected' | 'synced';

/** Payment methods */
export type PaymentMethod = 'credit_card' | 'debit_card' | 'cash' | 'check' | 'bank_transfer' | 'other';

/** Expense record */
export interface Expense {
  id: string;
  vendor_name: string;
  amount: number;
  currency: string;
  date: string;
  category: ExpenseCategory;
  payment_method: PaymentMethod;
  description: string;
  employee_id: string | null;
  is_reimbursable: boolean;
  tags: string[];
  status: ExpenseStatus;
  receipt_url: string | null;
  qbo_id: string | null;
  created_at: string;
  updated_at: string;
}

/** Receipt upload request */
export interface UploadReceiptRequest {
  receipt_data: string;  // Base64 encoded image
  content_type?: string; // image/png, image/jpeg, application/pdf
  employee_id?: string;
  payment_method?: PaymentMethod;
}

/** Create expense request */
export interface CreateExpenseRequest {
  vendor_name: string;
  amount: number;
  date?: string;
  category?: ExpenseCategory;
  payment_method?: PaymentMethod;
  description?: string;
  employee_id?: string;
  is_reimbursable?: boolean;
  tags?: string[];
}

/** Update expense request */
export interface UpdateExpenseRequest {
  vendor_name?: string;
  amount?: number;
  category?: ExpenseCategory;
  description?: string;
  status?: ExpenseStatus;
  is_reimbursable?: boolean;
  tags?: string[];
}

/** List expenses parameters */
export interface ListExpensesParams extends PaginationParams {
  category?: ExpenseCategory;
  vendor?: string;
  start_date?: string;
  end_date?: string;
  status?: ExpenseStatus;
  employee_id?: string;
}

/** Expense statistics */
export interface ExpenseStats {
  total_amount: number;
  expense_count: number;
  by_category: Record<ExpenseCategory, number>;
  by_status: Record<ExpenseStatus, number>;
  top_vendors: Array<{ vendor: string; amount: number; count: number }>;
  monthly_trend: Array<{ month: string; amount: number }>;
}

/** QBO sync result */
export interface SyncResult {
  success_count: number;
  failed_count: number;
  synced_ids: string[];
  errors: Array<{ expense_id: string; error: string }>;
}

/** Categorization result */
export interface CategorizeResult {
  categorized: Record<string, ExpenseCategory>;
  count: number;
}

// =============================================================================
// Expenses API
// =============================================================================

/**
 * Client interface for expenses operations.
 */
interface ExpensesClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Expenses namespace API for expense tracking and management.
 *
 * Provides comprehensive expense management:
 * - Receipt upload and OCR processing
 * - Auto-categorization using AI
 * - Approval workflows
 * - QuickBooks Online sync
 */
export class ExpensesAPI {
  constructor(private client: ExpensesClientInterface) {}

  // ===========================================================================
  // Receipt Processing
  // ===========================================================================

  /**
   * Upload and process a receipt image.
   *
   * Extracts vendor, amount, date, and category from the receipt
   * using OCR and AI analysis.
   *
   * @param request - Receipt data and metadata
   * @returns Created expense from receipt
   */
  async uploadReceipt(request: UploadReceiptRequest): Promise<{ expense: Expense; message: string }> {
    return this.client.request('POST', '/accounting/expenses/upload', {
      json: request as unknown as Record<string, unknown>,
    });
  }

  // ===========================================================================
  // CRUD Operations
  // ===========================================================================

  /**
   * Create an expense manually.
   *
   * @param request - Expense details
   * @returns Created expense
   */
  async create(request: CreateExpenseRequest): Promise<{ expense: Expense; message: string }> {
    return this.client.request('POST', '/accounting/expenses', {
      json: request as unknown as Record<string, unknown>,
    });
  }

  /**
   * List expenses with filters.
   *
   * @param params - Filtering and pagination options
   * @returns Paginated expense list
   */
  async list(params?: ListExpensesParams): Promise<{
    expenses: Expense[];
    total: number;
    limit: number;
    offset: number;
  }> {
    return this.client.request('GET', '/accounting/expenses', {
      params: params as unknown as Record<string, unknown>,
    });
  }

  /**
   * Get expense by ID.
   *
   * @param expenseId - Expense ID
   * @returns Expense details
   */
  async get(expenseId: string): Promise<{ expense: Expense }> {
    return this.client.request('GET', `/accounting/expenses/${expenseId}`);
  }

  /**
   * Update an expense.
   *
   * @param expenseId - Expense ID
   * @param request - Fields to update
   * @returns Updated expense
   */
  async update(expenseId: string, request: UpdateExpenseRequest): Promise<{ expense: Expense; message: string }> {
    return this.client.request('PUT', `/accounting/expenses/${expenseId}`, {
      json: request as unknown as Record<string, unknown>,
    });
  }

  /**
   * Delete an expense.
   *
   * @param expenseId - Expense ID
   * @returns Confirmation message
   */
  async delete(expenseId: string): Promise<{ message: string }> {
    return this.client.request('DELETE', `/accounting/expenses/${expenseId}`);
  }

  // ===========================================================================
  // Approval Workflow
  // ===========================================================================

  /**
   * Approve an expense for sync.
   *
   * @param expenseId - Expense ID
   * @returns Approved expense
   */
  async approve(expenseId: string): Promise<{ expense: Expense; message: string }> {
    return this.client.request('POST', `/accounting/expenses/${expenseId}/approve`);
  }

  /**
   * Reject an expense.
   *
   * @param expenseId - Expense ID
   * @param reason - Optional rejection reason
   * @returns Rejected expense
   */
  async reject(expenseId: string, reason?: string): Promise<{ expense: Expense; message: string }> {
    return this.client.request('POST', `/accounting/expenses/${expenseId}/reject`, {
      json: { reason },
    });
  }

  /**
   * Get expenses pending approval.
   *
   * @returns Pending expenses
   */
  async getPending(): Promise<{ expenses: Expense[]; count: number }> {
    return this.client.request('GET', '/accounting/expenses/pending');
  }

  // ===========================================================================
  // Categorization
  // ===========================================================================

  /**
   * Auto-categorize expenses using AI.
   *
   * @param expenseIds - Optional list of expense IDs (categorizes all uncategorized if empty)
   * @returns Categorization results
   */
  async categorize(expenseIds?: string[]): Promise<CategorizeResult & { message: string }> {
    return this.client.request('POST', '/accounting/expenses/categorize', {
      json: { expense_ids: expenseIds },
    });
  }

  // ===========================================================================
  // QBO Sync
  // ===========================================================================

  /**
   * Sync expenses to QuickBooks Online.
   *
   * @param expenseIds - Optional list of expense IDs (syncs all approved if empty)
   * @returns Sync results
   */
  async syncToQBO(expenseIds?: string[]): Promise<{ result: SyncResult; message: string }> {
    return this.client.request('POST', '/accounting/expenses/sync', {
      json: { expense_ids: expenseIds },
    });
  }

  // ===========================================================================
  // Statistics and Export
  // ===========================================================================

  /**
   * Get expense statistics.
   *
   * @param startDate - Optional start date filter
   * @param endDate - Optional end date filter
   * @returns Expense statistics
   */
  async getStats(startDate?: string, endDate?: string): Promise<{ stats: ExpenseStats }> {
    const params: Record<string, string> = {};
    if (startDate) params.start_date = startDate;
    if (endDate) params.end_date = endDate;
    return this.client.request('GET', '/accounting/expenses/stats', { params });
  }

  /**
   * Export expenses to CSV or JSON.
   *
   * @param format - Export format ('csv' or 'json')
   * @param startDate - Optional start date filter
   * @param endDate - Optional end date filter
   * @returns Exported data
   */
  async export(format: 'csv' | 'json' = 'csv', startDate?: string, endDate?: string): Promise<{ data: string; format: string }> {
    const params: Record<string, string> = { format };
    if (startDate) params.start_date = startDate;
    if (endDate) params.end_date = endDate;
    return this.client.request('GET', '/accounting/expenses/export', { params });
  }
}

File: sdk/python/aragora/namespaces/expenses.py
"""
Expenses Namespace API.

Provides a namespaced interface for expense tracking and management.
Supports receipt processing, auto-categorization, approval workflows, and QBO sync.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

ExpenseCategory = Literal[
    "travel",
    "meals",
    "office_supplies",
    "software",
    "equipment",
    "professional_services",
    "marketing",
    "utilities",
    "rent",
    "insurance",
    "other",
]
ExpenseStatus = Literal["pending", "approved", "rejected", "synced"]
PaymentMethod = Literal["credit_card", "debit_card", "cash", "check", "bank_transfer", "other"]


class ExpensesAPI:
    """
    Synchronous Expenses API.

    Provides methods for:
    - Receipt upload and OCR processing
    - Expense creation and management
    - Approval workflows
    - Auto-categorization
    - QuickBooks Online sync
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # =========================================================================
    # Receipt Processing
    # =========================================================================

    def upload_receipt(
        self,
        receipt_data: str,
        content_type: str | None = None,
        employee_id: str | None = None,
        payment_method: PaymentMethod | None = None,
    ) -> dict[str, Any]:
        """
        Upload and process a receipt image.

        Extracts vendor, amount, date, and category from the receipt
        using OCR and AI analysis.

        Args:
            receipt_data: Base64 encoded receipt image.
            content_type: MIME type (image/png, image/jpeg, application/pdf).
            employee_id: Employee identifier.
            payment_method: Payment method used.

        Returns:
            Created expense from receipt.
        """
        data: dict[str, Any] = {"receipt_data": receipt_data}
        if content_type:
            data["content_type"] = content_type
        if employee_id:
            data["employee_id"] = employee_id
        if payment_method:
            data["payment_method"] = payment_method

        return self._client.request("POST", "/api/v1/accounting/expenses/upload", json=data)

    # =========================================================================
    # CRUD Operations
    # =========================================================================

    def create(
        self,
        vendor_name: str,
        amount: float,
        date: str | None = None,
        category: ExpenseCategory | None = None,
        payment_method: PaymentMethod | None = None,
        description: str | None = None,
        employee_id: str | None = None,
        is_reimbursable: bool = False,
        tags: list[str] | None = None,
    ) -> dict[str, Any]:
        """
        Create an expense manually.

        Args:
            vendor_name: Vendor name.
            amount: Expense amount.
            date: Expense date (ISO format).
            category: Expense category.
            payment_method: Payment method used.
            description: Expense description.
            employee_id: Employee identifier.
            is_reimbursable: Whether expense is reimbursable.
            tags: Optional tags.

        Returns:
            Created expense.
        """
        data: dict[str, Any] = {"vendor_name": vendor_name, "amount": amount}
        if date:
            data["date"] = date
        if category:
            data["category"] = category
        if payment_method:
            data["payment_method"] = payment_method
        if description:
            data["description"] = description
        if employee_id:
            data["employee_id"] = employee_id
        if is_reimbursable:
            data["is_reimbursable"] = is_reimbursable
        if tags:
            data["tags"] = tags

        return self._client.request("POST", "/api/v1/accounting/expenses", json=data)

    def list(
        self,
        category: ExpenseCategory | None = None,
        vendor: str | None = None,
        start_date: str | None = None,
        end_date: str | None = None,
        status: ExpenseStatus | None = None,
        employee_id: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List expenses with filters.

        Args:
            category: Filter by category.
            vendor: Filter by vendor.
            start_date: Filter from date.
            end_date: Filter to date.
            status: Filter by status.
            employee_id: Filter by employee.
            limit: Maximum results.
            offset: Pagination offset.

        Returns:
            Paginated expense list.
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if category:
            params["category"] = category
        if vendor:
            params["vendor"] = vendor
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        if status:
            params["status"] = status
        if employee_id:
            params["employee_id"] = employee_id

        return self._client.request("GET", "/api/v1/accounting/expenses", params=params)

    def get(self, expense_id: str) -> dict[str, Any]:
        """
        Get expense by ID.

        Args:
            expense_id: Expense identifier.

        Returns:
            Expense details.
        """
        return self._client.request("GET", f"/api/v1/accounting/expenses/{expense_id}")

    def update(
        self,
        expense_id: str,
        vendor_name: str | None = None,
        amount: float | None = None,
        category: ExpenseCategory | None = None,
        description: str | None = None,
        status: ExpenseStatus | None = None,
        is_reimbursable: bool | None = None,
        tags: list[str] | None = None,
    ) -> dict[str, Any]:
        """
        Update an expense.

        Args:
            expense_id: Expense identifier.
            vendor_name: New vendor name.
            amount: New amount.
            category: New category.
            description: New description.
            status: New status.
            is_reimbursable: Update reimbursable flag.
            tags: New tags.

        Returns:
            Updated expense.
        """
        data: dict[str, Any] = {}
        if vendor_name:
            data["vendor_name"] = vendor_name
        if amount is not None:
            data["amount"] = amount
        if category:
            data["category"] = category
        if description:
            data["description"] = description
        if status:
            data["status"] = status
        if is_reimbursable is not None:
            data["is_reimbursable"] = is_reimbursable
        if tags is not None:
            data["tags"] = tags

        return self._client.request("PUT", f"/api/v1/accounting/expenses/{expense_id}", json=data)

    def delete(self, expense_id: str) -> dict[str, Any]:
        """
        Delete an expense.

        Args:
            expense_id: Expense identifier.

        Returns:
            Confirmation message.
        """
        return self._client.request("DELETE", f"/api/v1/accounting/expenses/{expense_id}")

    # =========================================================================
    # Approval Workflow
    # =========================================================================

    def approve(self, expense_id: str) -> dict[str, Any]:
        """
        Approve an expense for sync.

        Args:
            expense_id: Expense identifier.

        Returns:
            Approved expense.
        """
        return self._client.request("POST", f"/api/v1/accounting/expenses/{expense_id}/approve")

    def reject(
        self,
        expense_id: str,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """
        Reject an expense.

        Args:
            expense_id: Expense identifier.
            reason: Optional rejection reason.

        Returns:
            Rejected expense.
        """
        data: dict[str, Any] = {}
        if reason:
            data["reason"] = reason

        return self._client.request(
            "POST", f"/api/v1/accounting/expenses/{expense_id}/reject", json=data
        )

    def get_pending(self) -> dict[str, Any]:
        """
        Get expenses pending approval.

        Returns:
            Pending expenses with count.
        """
        return self._client.request("GET", "/api/v1/accounting/expenses/pending")

    # =========================================================================
    # Categorization
    # =========================================================================

    def categorize(
        self,
        expense_ids: list[str] | None = None,
    ) -> dict[str, Any]:
        """
        Auto-categorize expenses using AI.

        Args:
            expense_ids: Optional list of expense IDs (categorizes all uncategorized if empty).

        Returns:
            Categorization results.
        """
        data: dict[str, Any] = {}
        if expense_ids:
            data["expense_ids"] = expense_ids

        return self._client.request("POST", "/api/v1/accounting/expenses/categorize", json=data)

    # =========================================================================
    # QBO Sync
    # =========================================================================

    def sync_to_qbo(
        self,
        expense_ids: list[str] | None = None,
    ) -> dict[str, Any]:
        """
        Sync expenses to QuickBooks Online.

        Args:
            expense_ids: Optional list of expense IDs (syncs all approved if empty).

        Returns:
            Sync results.
        """
        data: dict[str, Any] = {}
        if expense_ids:
            data["expense_ids"] = expense_ids

        return self._client.request("POST", "/api/v1/accounting/expenses/sync", json=data)

    # =========================================================================
    # Statistics and Export
    # =========================================================================

    def get_stats(
        self,
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> dict[str, Any]:
        """
        Get expense statistics.

        Args:
            start_date: Optional start date filter.
            end_date: Optional end date filter.

        Returns:
            Expense statistics.
        """
        params: dict[str, Any] = {}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date

        return self._client.request("GET", "/api/v1/accounting/expenses/stats", params=params)

    def export(
        self,
        format: str = "csv",
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> dict[str, Any]:
        """
        Export expenses to CSV or JSON.

        Args:
            format: Export format ('csv' or 'json').
            start_date: Optional start date filter.
            end_date: Optional end date filter.

        Returns:
            Exported data.
        """
        params: dict[str, Any] = {"format": format}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date

        return self._client.request("GET", "/api/v1/accounting/expenses/export", params=params)


class AsyncExpensesAPI:
    """Asynchronous Expenses API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    # =========================================================================
    # Receipt Processing
    # =========================================================================

    async def upload_receipt(
        self,
        receipt_data: str,
        content_type: str | None = None,
        employee_id: str | None = None,
        payment_method: PaymentMethod | None = None,
    ) -> dict[str, Any]:
        """Upload and process a receipt image."""
        data: dict[str, Any] = {"receipt_data": receipt_data}
        if content_type:
            data["content_type"] = content_type
        if employee_id:
            data["employee_id"] = employee_id
        if payment_method:
            data["payment_method"] = payment_method

        return await self._client.request("POST", "/api/v1/accounting/expenses/upload", json=data)

    # =========================================================================
    # CRUD Operations
    # =========================================================================

    async def create(
        self,
        vendor_name: str,
        amount: float,
        date: str | None = None,
        category: ExpenseCategory | None = None,
        payment_method: PaymentMethod | None = None,
        description: str | None = None,
        employee_id: str | None = None,
        is_reimbursable: bool = False,
        tags: list[str] | None = None,
    ) -> dict[str, Any]:
        """Create an expense manually."""
        data: dict[str, Any] = {"vendor_name": vendor_name, "amount": amount}
        if date:
            data["date"] = date
        if category:
            data["category"] = category
        if payment_method:
            data["payment_method"] = payment_method
        if description:
            data["description"] = description
        if employee_id:
            data["employee_id"] = employee_id
        if is_reimbursable:
            data["is_reimbursable"] = is_reimbursable
        if tags:
            data["tags"] = tags

        return await self._client.request("POST", "/api/v1/accounting/expenses", json=data)

    async def list(
        self,
        category: ExpenseCategory | None = None,
        vendor: str | None = None,
        start_date: str | None = None,
        end_date: str | None = None,
        status: ExpenseStatus | None = None,
        employee_id: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List expenses with filters."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if category:
            params["category"] = category
        if vendor:
            params["vendor"] = vendor
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        if status:
            params["status"] = status
        if employee_id:
            params["employee_id"] = employee_id

        return await self._client.request("GET", "/api/v1/accounting/expenses", params=params)

    async def get(self, expense_id: str) -> dict[str, Any]:
        """Get expense by ID."""
        return await self._client.request("GET", f"/api/v1/accounting/expenses/{expense_id}")

    async def update(
        self,
        expense_id: str,
        vendor_name: str | None = None,
        amount: float | None = None,
        category: ExpenseCategory | None = None,
        description: str | None = None,
        status: ExpenseStatus | None = None,
        is_reimbursable: bool | None = None,
        tags: list[str] | None = None,
    ) -> dict[str, Any]:
        """Update an expense."""
        data: dict[str, Any] = {}
        if vendor_name:
            data["vendor_name"] = vendor_name
        if amount is not None:
            data["amount"] = amount
        if category:
            data["category"] = category
        if description:
            data["description"] = description
        if status:
            data["status"] = status
        if is_reimbursable is not None:
            data["is_reimbursable"] = is_reimbursable
        if tags is not None:
            data["tags"] = tags

        return await self._client.request(
            "PUT", f"/api/v1/accounting/expenses/{expense_id}", json=data
        )

    async def delete(self, expense_id: str) -> dict[str, Any]:
        """Delete an expense."""
        return await self._client.request("DELETE", f"/api/v1/accounting/expenses/{expense_id}")

    # =========================================================================
    # Approval Workflow
    # =========================================================================

    async def approve(self, expense_id: str) -> dict[str, Any]:
        """Approve an expense for sync."""
        return await self._client.request(
            "POST", f"/api/v1/accounting/expenses/{expense_id}/approve"
        )

    async def reject(
        self,
        expense_id: str,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """Reject an expense."""
        data: dict[str, Any] = {}
        if reason:
            data["reason"] = reason

        return await self._client.request(
            "POST", f"/api/v1/accounting/expenses/{expense_id}/reject", json=data
        )

    async def get_pending(self) -> dict[str, Any]:
        """Get expenses pending approval."""
        return await self._client.request("GET", "/api/v1/accounting/expenses/pending")

    # =========================================================================
    # Categorization
    # =========================================================================

    async def categorize(
        self,
        expense_ids: list[str] | None = None,
    ) -> dict[str, Any]:
        """Auto-categorize expenses using AI."""
        data: dict[str, Any] = {}
        if expense_ids:
            data["expense_ids"] = expense_ids

        return await self._client.request(
            "POST", "/api/v1/accounting/expenses/categorize", json=data
        )

    # =========================================================================
    # QBO Sync
    # =========================================================================

    async def sync_to_qbo(
        self,
        expense_ids: list[str] | None = None,
    ) -> dict[str, Any]:
        """Sync expenses to QuickBooks Online."""
        data: dict[str, Any] = {}
        if expense_ids:
            data["expense_ids"] = expense_ids

        return await self._client.request("POST", "/api/v1/accounting/expenses/sync", json=data)

    # =========================================================================
    # Statistics and Export
    # =========================================================================

    async def get_stats(
        self,
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> dict[str, Any]:
        """Get expense statistics."""
        params: dict[str, Any] = {}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date

        return await self._client.request("GET", "/api/v1/accounting/expenses/stats", params=params)

    async def export(
        self,
        format: str = "csv",
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> dict[str, Any]:
        """Export expenses to CSV or JSON."""
        params: dict[str, Any] = {"format": format}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date

        return await self._client.request(
            "GET", "/api/v1/accounting/expenses/export", params=params
        )
